lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,315 @@
+   * @return {Promise<RootNode[]>}
+   */
+  static async makeJavaScriptRootNodes(artifacts, context) {
+    /** @type {RootNode[]} */
+    const rootNodes = [];
+
+    const bundles = await JsBundles.request(artifacts, context);
+    const duplication = await ModuleDuplication.request(artifacts, context);
+
+    let inlineScriptLength = 0;
+    for (const scriptElement of artifacts.ScriptElements) {
+      // No src means script is inline.
+      // Combine these ScriptElements so that inline scripts show up as a single root node.
+      if (!scriptElement.src) {
+        inlineScriptLength += (scriptElement.content || '').length;
+      }
+    }
+    if (inlineScriptLength) {
+      const name = artifacts.URL.finalUrl;
+      rootNodes.push({
+        name,
+        node: {
+          name,
+          resourceBytes: inlineScriptLength,
+          unusedBytes: 0,
+        },
+      });
+    }
+
+    for (const scriptElement of artifacts.ScriptElements) {
+      if (!scriptElement.src) continue;
+
+      const name = scriptElement.src;
+      const bundle = bundles.find(bundle => scriptElement.src === bundle.script.src);
+      const scriptCoverages = artifacts.JsUsage[scriptElement.src];
+      if (!bundle || !scriptCoverages) {
+        rootNodes.push({
+          name,
+          node: {
+            name,
+            resourceBytes: scriptElement.src.length,
+          },
+        });
+        continue;
+      }
+
+      const unusedJavascriptSummary = await UnusedJavaScriptSummary.request(
+        {url: scriptElement.src, scriptCoverages, bundle}, context);
+
+      let node;
+      if (!unusedJavascriptSummary.sourcesWastedBytes) {
+        node = {
+          name,
+          resourceBytes: unusedJavascriptSummary.totalBytes,
+          unusedBytes: unusedJavascriptSummary.wastedBytes,
+        };
+      } else {
+        /** @type {Record<string, SourceData>} */
+        const sourcesData = {};
+        for (const source of Object.keys(bundle.sizes.files)) {
+          /** @type {SourceData} */
+          const sourceData = {
+            resourceBytes: bundle.sizes.files[source],
+            unusedBytes: unusedJavascriptSummary.sourcesWastedBytes[source],
+          };
+
+          const key = ModuleDuplication.normalizeSource(source);
+          if (duplication.has(key)) sourceData.duplicate = key;
+
+          sourcesData[source] = sourceData;
+        }
+
+        node = this.prepareTreemapNodes(bundle.rawMap.sourceRoot || '', sourcesData);
+      }
+
+      rootNodes.push({
+        name,
+        node,
+      });
+    }
+
+    return rootNodes;
+  }
+
+  /**
+   * Returns root node where the first level is network resource type,
+   * followed by the leaf nodes which are are all network requests. Data
+   * contained is the resourceBytes.
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<RootNode>}
+   */
+  static async makeResourceSummaryRootNode(artifacts, context) {
```

Comment:
I'm not sure I get why this needs to be part of the treemap. Isn't this data already exposed in the LHR in other audits?

Reply:
You're right, this could be derived from `network-requests` audit.

I'd prefer to encapsulate all of the tree-making code in this one place, than do any of that in the treemap app. If the size is bigger than we'd like, I'm fine with moving this to runtime in the treemap app.

Reply:
That seems a little backwards to me on first blush. Shouldn't Lighthouse provide all of the base-level raw data in a format untethered to particular UIs and the treemap app be responsible for converting it into tree-like structures? 

For the bundle/module stuff this all makes sense because what other way are we going surface this info, I mean folder structures are natural trees and other consumers can still make use of it, but this resource summary fake tree thing is duplicating data we already have available elsewhere in a way that seems extremely specific and useful only to this treemap app.

Reply:
Now that I'm thinking about it I guess I'm asking why do we have this in a `treemap-data` audit instead of `bundle-module-structure` or something

Reply:
`script-treemap-data` ?